### PR TITLE
feat: Implement testing

### LINKED ISSUE #496: [REQUEST] Unit testing
**What is the feature you're proposing?**
`unstable` is a land of both creativity and instability, which is commonly seen from the consistency in bugs appearing on this branch. This branch helps us with standardising a rapid development workflow, however, we often find ourselves having issues with detecting when regular core features of the library begin to break. It's hard to always be able to check every area of this library ourselves, so, I would like to suggest unit testing. This would be a sound and dependable way for us to keep a consistent method in testing parts of the library and halting our ability from committing new code if something fails as a check.

The introduction of unit testing would work in a similar fashion to `pre-commit`. Additionally, we've now enabled readthedocs builds to occur per Pull Request made in this repository, which will be useful for documentation-based changes as well.

**Additional Information**
- [ ] My feature request is related to an existing Issue.
    - Issue (if referenceable):

### PR DESCRIPTION
**Pull request by sillyElvy**
## About

This pull request implements testing by creating objects and checking if the generated payloads are equal to the previously checked and correct payloads.

This includes eventual modifications of existing code to provide testing possibilities

## Checklist

- [x] The ``pre-commit`` code linter has been run over all edited files to ensure the code is linted.
- [x] I've ensured the change(s) work on `3.8.6` and higher.

I've made this pull request: (check all that apply)
  - [ ] For the documentation
  - [x] To add a new feature
  - [x] As a general enhancement
  - [ ] As a refactor of the library/the library's code
  - [ ] To fix an existing bug
  - [x] To finally resolve #496

This is:
  - [ ] A breaking change

### CODE REVIEW

File: tests/test_new_command_system.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    assert command.full_data == {
+        "default_member_permissions": "2147483648",
+        "description": "Hello!",
+        "description_localizations": {},
+        "dm_permission": True,
+        "name": "command",
+        "name_localizations": {},
+        "options": [],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]:
```suggestion
    assert command.full_data == dict(
        default_member_permissions="2147483648",
        description="Hello!",
        description_localizations={},
        dm_permission=True,
        name="command",
        name_localizations={},
        options=[],
        type=1,
    )
```
Looks nicer imo

File: tests/test_new_command_system.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        "options": [
+            {
+                "description": "hi",
+                "name": "option",
+                "required": True,
+                "type": interactions.OptionType.STRING,
+            }
+        ],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]: https://github.com/interactions-py/library/pull/1132#discussion_r1002181715

File: tests/test_new_command_system.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        "options": [
+            {
+                "description": "hi",
+                "name": "option",
+                "required": False,
+                "type": interactions.OptionType.STRING,
+            }
+        ],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]: https://github.com/interactions-py/library/pull/1132#discussion_r1002181715

File: tests/test_new_command_system.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        "options": [
+            {
+                "description": "hi",
+                "name": "option",
+                "required": True,
+                "type": interactions.OptionType.USER,
+            }
+        ],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]: https://github.com/interactions-py/library/pull/1132#discussion_r1002181715

File: tests/test_new_command_system.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        "options": [
+            {
+                "description": "hi",
+                "name": "option",
+                "required": False,
+                "type": interactions.OptionType.USER,
+            }
+        ],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]: https://github.com/interactions-py/library/pull/1132#discussion_r1002181715

File: tests/test_new_command_system.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+
+def test_basic_command_with_connector_option(fake_client):
+    @fake_client.command()
+    @interactions.option(
+        "hi", name="hi", type=interactions.OptionType.USER, required=True, connector="d"
```
Comment (Toricane) [inline tests/test_new_command_system.py:205]:
```suggestion
        "hi", name="hi", type=interactions.OptionType.USER, required=True
```
Connector is automatically set as parameter name if name= is specified

File: tests/test_new_command_system.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            {
+                "connector": "d",
+                "description": "hi",
+                "name": "hi",
+                "required": True,
+                "type": interactions.OptionType.USER,
+            }
+        ],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]: https://github.com/interactions-py/library/pull/1132#discussion_r1002181715

File: tests/test_new_command_system.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    assert command.full_data == {
+        "default_member_permissions": "8",
+        "description": "Hello!",
+        "description_localizations": {},
+        "dm_permission": False,
+        "name": "command",
+        "name_localizations": {},
+        "options": [],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]: https://github.com/interactions-py/library/pull/1132#discussion_r1002181715

File: tests/test_new_command_system.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+            {
+                "choices": [{"name": "hi", "value": "hi"}, {"name": "bye", "value": "bye"}],
+                "description": "hi",
+                "name": "option",
+                "required": False,
+                "type": interactions.OptionType.STRING,
+            }
+        ],
+        "type": 1,
+    }
```
Comment (Toricane) [inline tests/test_new_command_system.py]: https://github.com/interactions-py/library/pull/1132#discussion_r1002181715

**Review by Toricane — CHANGES_REQUESTED** [review]
Some suggestions

File: tests/test_new_command_system.py (hunk 6 above)
Comment (sillyElvy) [reply tests/test_new_command_system.py:205]: Oh cool then I have to make two test cases for it xD

Comment (Toricane) [reply tests/test_new_command_system.py:205]: Oh yeah btw required= is automatic too based on whether there is a default for the parameter

Comment (sillyElvy) [reply tests/test_new_command_system.py:205]: If you looked at the function names you'd know 😅

I'm checking if the overriding works (it's inverted to what it should be without override) 

If any of the checks fails then something is not ok

Comment (sillyElvy) [reply tests/test_new_command_system.py:205]: > Connector is automatically set as parameter name if name= is specified

uhh.. yeah so about that:
```py
============================================================================================================== FAILURES =============================================================================================================== 
___________________________________________________________________________________ test_basic_command_with_connector_option_with_read_from_kwargs ____________________________________________________________________________________ 

fake_client = <interactions.client.bot.Client object at 0x0000022169ADECE0>

    def test_basic_command_with_connector_option_with_read_from_kwargs(fake_client):
        @fake_client.command()
        @interactions.option("hi", name="hi")
        async def command(ctx, d: interactions.Member):
            """Hello!"""

>       assert command.full_data == dict(
            default_member_permissions="2147483648",
            description="Hello!",
            description_localizations={},
            dm_permission=True,
            name="command",
            name_localizations={},
            options=[
                {
                    "connector": "d",
                    "description": "hi",
                    "name": "hi",
                    "required": True,
                    "type": interactions.OptionType.USER,
                }
            ],
            type=1,
        )
E       AssertionError: assert {'default_mem...n': True, ...} == {'default_mem...n': True, ...}
E         Omitting 7 identical items, use -vv to show
E         Differing items:
E         {'options': [{'description': 'hi', 'name': 'hi', 'required': True, 'type': <OptionType.USER: 6>}]} != {'options': [{'connector': 'd', 'description': 'hi', 'name': 'hi', 'required': True, ...}]}
E         Use -v to get more diff

tests\test_new_command_system.py:131: AssertionError
======================================================================================================= short test summary info ======================================================================================================= 
FAILED tests/test_new_command_system.py::test_basic_command_with_connector_option_with_read_from_kwargs - AssertionError: assert {'default_mem...n': True, ...} == {'default_mem...n': True, ...}
```

Comment (sillyElvy) [reply tests/test_new_command_system.py:205]: oh I see how you did it, this is interesting

Comment (Toricane) [reply tests/test_new_command_system.py:205]: Wait what happened

Comment (sillyElvy) [reply tests/test_new_command_system.py:205]: Well the "connector" is shown in the option dict (check code above) if its put inside the option decorator but not if its made automatically and I thought that was a bug but command.connectors still is correct

File: tests/test_library_exception.py [hunk 10]
```diff
@@ -0,0 +1,860 @@
...
+    170004,
+    170005,
+    170006,
+    170007,
+    180000,
+    180002,
+    200000,
+    200001,
+    220003,
+)
```
Comment (Toricane) [inline tests/test_library_exception.py]:
```suggestion
codes = (
    *range(15),
    400,
    401,
    *range(403, 406),
    429,
    502,
    *range(4000, 4015),
    *range(10001, 10017),
    10020,
    *range(10026, 10034),
    10036,
    10038,
    10049,
    10050,
    10057,
    10059,
    10060,
    10062,
    10063,
    *range(10065, 10072),
    10087,
    20001,
    20002,
    20009,
    20012,
    20016,
    20018,
    20022,
    20024,
    20028,
    20029,
    20031,
    20035,
    *range(30001, 30006),
    30007,
    30008,
    30010,
    30013,
    30015,
    30016,
    30018,
    30019,
    *range(30030, 30036),
    *range(30037, 30041),
    30042,
    30046,
    30047,
    30048,
    30052,
    *range(40001, 40008),
    40012,
    40032,
    40033,
    40041,
    40043,
    40060,
    40061,
    *range(50001, 50018),
    *range(50019, 50022),
    *range(50024, 50029),
    *range(50033, 50037),
    50041,
    50045,
    50046,
    50054,
    50055,
    50068,
    50070,
    50074,
    50080,
    50081,
    *range(50083, 50087),
    50095,
    50097,
    50101,
    50109,
    50132,
    50138,
    50146,
    50600,
    60003,
    80004,
    90001,
    110001,
    130000,
    150006,
    160002,
    *range(160004, 160008),
    *range(170001, 170008),
    180000,
    180002,
    200000,
    200001,
    220003,
)
```
i spent way too long on this

File: tests/test_library_exception.py [hunk 11]
```diff
@@ -0,0 +1,860 @@
...
+                                "message": "This field is required",
+                            }
+                        ]
+                    },
+                }
+            }
+        }
+    },
+    "message": "Invalid Form Body",
+}
```
Comment (Toricane) [inline tests/test_library_exception.py]:
```suggestion
__error = dict(
    _errors=[
        dict(
            code="BASE_TYPE_REQUIRED",
            message="This field is required",
        )
    ]
)
data = dict(
    code=50035,
    errors=dict(
        components={
            "0": dict(
                components={
                    **{str(i): __error for i in range(10)},
                    **{str(i): __error for i in range(10, 74)},
                }
            )
        }
    ),
    message="Invalid Form Body",
)
```

File: tests/test_new_command_system.py [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+
+    assert command.full_data == dict(
+        default_member_permissions="2147483648",
+        description="Hello!",
+        description_localizations={},
+        dm_permission=True,
+        name="command",
+        name_localizations={},
+        options=[
+            {
```
Comment (Toricane) [inline tests/test_new_command_system.py:57]: Convert all inner dicts to `dict()`?

**Review by Toricane — CHANGES_REQUESTED** [review]
So much unnecessary code

File: tests/test_library_exception.py (hunk 10 above)
Comment (sillyElvy) [reply tests/test_library_exception.py]: XD

Comment (Toricane) [reply tests/test_library_exception.py]: can you commit this again? i removed some underscores

Comment (Toricane) [reply tests/test_library_exception.py]: @EdVraz

File: tests/test_new_command_system.py (hunk 12 above)
Comment (Toricane) [reply tests/test_new_command_system.py:57]: Theres a lot of these

File: tests/test_library_exception.py (hunk 10 above)
Comment (sillyElvy) [reply tests/test_library_exception.py]: "comment is outdated" you have to make a new one

File: tests/test_new_command_system.py (hunk 12 above)
Comment (sillyElvy) [reply tests/test_new_command_system.py:57]: Why tho, I understand that it's nicer for the total but for the few options in total I don't think it's nessecary

File: tests/test_library_exception.py [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    *range(15),
+    400,
+    401,
+    *range(403, 406),
+    429,
+    502,
+    *range(4000, 4015),
+    *range(10_001, 10_017),
+    10_020,
+    *range(10_026, 10_034),
```
Comment (Toricane) [inline tests/test_library_exception.py]:
```suggestion
    *range(10001, 10017),
    10020,
    *range(10026, 10034),
```

File: tests/test_library_exception.py [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    10036,
+    10038,
+    10049,
+    10050,
+    10057,
+    10059,
+    10060,
+    10062,
+    10063,
+    *range(10_065, 10_072),
```
Comment (Toricane) [inline tests/test_library_exception.py]:
```suggestion
    *range(10065, 10072),
```

File: tests/test_library_exception.py [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    20012,
+    20016,
+    20018,
+    20022,
+    20024,
+    20028,
+    20029,
+    20031,
+    20035,
+    *range(30_001, 30_006),
```
Comment (Toricane) [inline tests/test_library_exception.py]:
```suggestion
    *range(30001, 30006),
```

**Review by Toricane — CHANGES_REQUESTED** [review]

File: tests/test_new_command_system.py (hunk 12 above)
Comment (Toricane) [reply tests/test_new_command_system.py:57]: fine

**Pull request closed without merging**